File: src/interactive-preview.ts
import { randomBytes } from "node:crypto";
import { open, realpath } from "node:fs/promises";
import { basename, dirname, extname, isAbsolute, join, relative, resolve } from "node:path";
import { AuthStore, parseCookie, SESSION_COOKIE } from "./auth.ts";
import { MAX_DOCUMENT_BYTES, MAX_PREVIEW_IMAGE_BYTES, previewType, withinPreviewRoot } from "./preview-security.ts";

export const INTERACTIVE_CSP = "default-src 'none'; script-src 'unsafe-inline' 'unsafe-eval' http: https: blob:; style-src 'unsafe-inline' http: https:; img-src http: https: data: blob:; font-src http: https: data:; media-src http: https: data: blob:; connect-src http: https: ws: wss:; worker-src blob:; frame-src 'none'; object-src 'none'; sandbox allow-scripts; base-uri 'none'; form-action 'none'; frame-ancestors 'self'";

const ASSET_TYPES: Record<string, string> = {
  ".js": "text/javascript; charset=utf-8", ".mjs": "text/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8", ".json": "application/json", ".csv": "text/csv; charset=utf-8",
  ".txt": "text/plain; charset=utf-8", ".wasm": "application/wasm",
  ".woff": "font/woff", ".woff2": "font/woff2", ".ttf": "font/ttf", ".otf": "font/otf",
  ".ico": "image/x-icon", ".avif": "image/avif", ".bmp": "image/bmp",
  ".mp3": "audio/mpeg", ".wav": "audio/wav", ".ogg": "audio/ogg", ".mp4": "video/mp4", ".webm": "video/webm",
};

type Grant = { directory: string; session: string; expires: number };

export class InteractivePreviews {
  private readonly grants = new Map<string, Grant>();

  constructor(private auth: AuthStore, private roots: readonly string[], private stateDirectory: string) {}

  async create(path: string, cookie: string | null): Promise<{ token: string; url: string } | null> {
    const session = parseCookie(cookie, SESSION_COOKIE);
    if (!session || !this.auth.authenticated(cookie) || !isAbsolute(path)) return null;
    try {
      const canonical = await realpath(path);
      if (previewType(canonical)?.kind !== "html" || !withinPreviewRoot(canonical, this.roots)) return null;
      const directory = dirname(canonical);
      const grant = { directory, session, expires: Date.now() + 24 * 60 * 60 * 1000 };
      if (!(await this.readFile(grant, basename(canonical)))) return null;
      for (const [key, entry] of this.grants) if (!this.valid(entry)) this.grants.delete(key);
      const token = randomBytes(32).toString("hex");
      this.grants.set(token, grant);
      return { token, url: `/api/preview-content/${token}/${encodeURIComponent(basename(canonical))}` };
    } catch { return null; }
  }

  revoke(token: string, cookie: string | null): boolean {
    if (this.grants.get(token)?.session !== parseCookie(cookie, SESSION_COOKIE)) return false;
    return this.grants.delete(token);
  }

  async read(token: string, resource: string): Promise<{ bytes: Uint8Array<ArrayBuffer>; contentType: string } | null> {
    const grant = this.grants.get(token);
    if (!grant || !this.valid(grant)) { this.grants.delete(token); return null; }
    return this.readFile(grant, resource);
  }

  private valid(grant: Grant): boolean {
    return grant.expires > Date.now() && this.auth.authenticated(`${SESSION_COOKIE}=${grant.session}`);
  }

  private async readFile(grant: Grant, resource: string): Promise<{ bytes: Uint8Array<ArrayBuffer>; contentType: string } | null> {
    if (!resource || isAbsolute(resource) || resource.includes("\\") || resource.includes("\0") || resource.split("/").some(part => part.startsWith("."))) return null;
    try {
      const canonical = await realpath(join(grant.directory, resource));
      const privateDirectory = await realpath(this.stateDirectory).catch(() => resolve(this.stateDirectory));
      if (!withinPreviewRoot(canonical, [grant.directory]) || !withinPreviewRoot(canonical, this.roots) || withinPreviewRoot(canonical, [privateDirectory]) || relative(grant.directory, canonical).split("/").some(part => part.startsWith("."))) return null;
      const kind = previewType(canonical);
      const contentType = kind?.contentType ?? ASSET_TYPES[extname(canonical).toLowerCase()];
      if (!contentType || kind?.kind === "markdown") return null;
      const maxBytes = kind?.maxBytes ?? (contentType.startsWith("text/") || contentType === "application/json" ? MAX_DOCUMENT_BYTES : MAX_PREVIEW_IMAGE_BYTES);
      const file = await open(canonical, "r");
      try {
        const metadata = await file.stat();
        if (!metadata.isFile() || metadata.size > maxBytes) return null;
        const bytes = Buffer.alloc(maxBytes + 1);
        let length = 0;
        while (length < bytes.length) {
          const result = await file.read(bytes, length, bytes.length - length, null);
          if (!result.bytesRead) break;
          length += result.bytesRead;
        }
        return length <= maxBytes ? { bytes: bytes.subarray(0, length), contentType } : null;
      } finally { await file.close(); }
    } catch { return null; }
  }
}

File: rust/interactive_preview.rs
use super::*;
use rand_core::{OsRng, RngCore};
use std::collections::HashMap;

pub const INTERACTIVE_CSP: &str = "default-src 'none'; script-src 'unsafe-inline' 'unsafe-eval' http: https: blob:; style-src 'unsafe-inline' http: https:; img-src http: https: data: blob:; font-src http: https: data:; media-src http: https: data: blob:; connect-src http: https: ws: wss:; worker-src blob:; frame-src 'none'; object-src 'none'; sandbox allow-scripts; base-uri 'none'; form-action 'none'; frame-ancestors 'self'";

#[derive(Clone)]
struct Grant {
    directory: PathBuf,
    session: String,
    expires: SystemTime,
}

#[derive(Clone, Default)]
pub struct InteractivePreviews(Arc<Mutex<HashMap<String, Grant>>>);

impl InteractivePreviews {
    fn valid(grant: &Grant, auth: &AuthStore) -> bool {
        grant.expires > SystemTime::now()
            && auth.authenticated(Some(&format!("{}={}", auth::SESSION_COOKIE, grant.session)))
    }
}

pub async fn create(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(query): Json<PreviewQuery>,
) -> Response {
    if !allowed_host(&state, &headers) || !allowed_origin(&state, &headers) {
        return secure((StatusCode::FORBIDDEN, "Cross-origin rejected").into_response());
    }
    if !authenticated(&state, &headers) {
        return secure((StatusCode::UNAUTHORIZED, "Authentication required").into_response());
    }
    let requested = Path::new(&query.path);
    if !requested.is_absolute() {
        return secure((StatusCode::BAD_REQUEST, "Absolute path required").into_response());
    }
    let Ok(path) = fs::canonicalize(requested) else {
        return unavailable();
    };
    if !within_preview_root(&path, &state.preview_roots)
        || preview_type(&path).is_none_or(|kind| kind.kind != "html")
    {
        return unavailable();
    }
    let session = cookie_header(&headers)
        .and_then(|cookie| auth::parse_cookie(cookie, auth::SESSION_COOKIE))
        .unwrap()
        .to_string();
    let grant = Grant {
        directory: path.parent().unwrap().to_path_buf(),
        session,
        expires: SystemTime::now() + Duration::from_secs(24 * 60 * 60),
    };
    let Some(name) = path.file_name().and_then(|name| name.to_str()) else {
        return unavailable();
    };
    if read_file(&state, &grant, name).is_none() {
        return unavailable();
    }
    let mut random = [0u8; 32];
    OsRng.fill_bytes(&mut random);
    let token: String = random.iter().map(|byte| format!("{byte:02x}")).collect();
    let url = format!(
        "/api/preview-content/{token}/{}",
        percent_encode(name.as_bytes())
    );
    let mut grants = state.interactive_previews.0.lock().unwrap();
    grants.retain(|_, grant| InteractivePreviews::valid(grant, &state.auth));
    grants.insert(token.clone(), grant);
    let mut response =
        secure(Json(serde_json::json!({"token": token, "url": url})).into_response());
    response
        .headers_mut()
        .insert(header::CACHE_CONTROL, HeaderValue::from_static("no-store"));
    response
}

pub async fn revoke(
    State(state): State<AppState>,
    headers: HeaderMap,
    axum::extract::Path(token): axum::extract::Path<String>,
) -> Response {
    if !allowed_host(&state, &headers) || !allowed_origin(&state, &headers) {
        return secure((StatusCode::FORBIDDEN, "Cross-origin rejected").into_response());
    }
    if !authenticated(&state, &headers) {
        return secure((StatusCode::UNAUTHORIZED, "Authentication required").into_response());
    }
    let session =
        cookie_header(&headers).and_then(|cookie| auth::parse_cookie(cookie, auth::SESSION_COOKIE));
    let mut grants = state.interactive_previews.0.lock().unwrap();
    if grants
        .get(&token)
        .is_some_and(|grant| Some(grant.session.as_str()) == session)
    {
        grants.remove(&token);
    }
    secure(StatusCode::NO_CONTENT.into_response())
}

pub async fn content(
    State(state): State<AppState>,
    headers: HeaderMap,
    axum::extract::Path((token, resource)): axum::extract::Path<(String, String)>,
) -> Response {
    if !allowed_host(&state, &headers) {
        return secure((StatusCode::FORBIDDEN, "Forbidden host").into_response());
    }
    let grant = state
        .interactive_previews
        .0
        .lock()
        .unwrap()
        .get(&token)
        .cloned();
    let Some(grant) = grant else {
        return unavailable();
    };
    if !InteractivePreviews::valid(&grant, &state.auth) {
        state.interactive_previews.0.lock().unwrap().remove(&token);
        return unavailable();
    }
    let Some((bytes, content_type)) = read_file(&state, &grant, &resource) else {
        return unavailable();
    };
    let mut response = bytes.into_response();
    let headers = response.headers_mut();
    headers.insert(header::CONTENT_TYPE, HeaderValue::from_static(content_type));
    headers.insert(header::CACHE_CONTROL, HeaderValue::from_static("no-store"));
    headers.insert(
        header::CONTENT_DISPOSITION,
        HeaderValue::from_static("inline"),
    );
    headers.insert(
        header::CONTENT_SECURITY_POLICY,
        HeaderValue::from_static(INTERACTIVE_CSP),
    );
    headers.insert(
        header::ACCESS_CONTROL_ALLOW_ORIGIN,
        HeaderValue::from_static("*"),
    );
    headers.insert(
        "x-content-type-options",
        HeaderValue::from_static("nosniff"),
    );
    headers.insert("referrer-policy", HeaderValue::from_static("no-referrer"));
    response
}

fn unavailable() -> Response {
    let mut response =
        secure((StatusCode::NOT_FOUND, "预览资源不可用，请刷新预览页").into_response());
    response
        .headers_mut()
        .insert(header::CACHE_CONTROL, HeaderValue::from_static("no-store"));
    response
}

fn read_file(state: &AppState, grant: &Grant, resource: &str) -> Option<(Vec<u8>, &'static str)> {
    if resource.is_empty()
        || Path::new(resource).is_absolute()
        || resource.contains(['\\', '\0'])
        || resource.split('/').any(|part| part.starts_with('.'))
    {
        return None;
    }
    let canonical = fs::canonicalize(grant.directory.join(resource)).ok()?;
    let private_directory = fs::canonicalize(state_dir()).ok()?;
    if !canonical.starts_with(&grant.directory)
        || !within_preview_root(&canonical, &state.preview_roots)
        || canonical.starts_with(private_directory)
        || canonical
            .strip_prefix(&grant.directory)
            .ok()?
            .components()
            .any(|part| part.as_os_str().to_string_lossy().starts_with('.'))
    {
        return None;
    }
    let (content_type, max_bytes) = if let Some(kind) = preview_type(&canonical) {
        if kind.kind == "markdown" {
            return None;
        }
        (kind.content_type, kind.max_bytes)
    } else {
        let extension = canonical.extension()?.to_str()?.to_ascii_lowercase();
        let content_type = match extension.as_str() {
            "js" | "mjs" => "text/javascript; charset=utf-8",
            "css" => "text/css; charset=utf-8",
            "json" => "application/json",
            "csv" => "text/csv; charset=utf-8",
            "txt" => "text/plain; charset=utf-8",
            "wasm" => "application/wasm",
            "woff" => "font/woff",
            "woff2" => "font/woff2",
            "ttf" => "font/ttf",
            "otf" => "font/otf",
            "ico" => "image/x-icon",
            "avif" => "image/avif",
            "bmp" => "image/bmp",
            "mp3" => "audio/mpeg",
            "wav" => "audio/wav",
            "ogg" => "audio/ogg",
            "mp4" => "video/mp4",
            "webm" => "video/webm",
            _ => return None,
        };
        let max_bytes = if content_type.starts_with("text/") || content_type == "application/json" {
            MAX_DOCUMENT_BYTES
        } else {
            MAX_PREVIEW_IMAGE_BYTES
        };
        (content_type, max_bytes)
    };
    let mut file = fs::File::open(canonical).ok()?;
    let metadata = file.metadata().ok()?;
    if !metadata.is_file() || metadata.len() > max_bytes {
        return None;
    }
    let mut bytes = Vec::new();
    (&mut file)
        .take(max_bytes + 1)
        .read_to_end(&mut bytes)
        .ok()?;
    (bytes.len() as u64 <= max_bytes).then_some((bytes, content_type))
}
